### PR TITLE
Refactoring/review 300123

### PR DESCRIPTION
**Pull request by mithunbharadwaj**
**What is in PR?**

- [x] each orchestrator has its own specific Template check.
- [x] Render API now works.
- [x] library restructuring
- [x] Rename mail to email everywhere
- [x] Extend qa.md with other tests (see below)
- [x] Replace asserts for input param validations (other exception needs to be raised)
- [ ] `/Templates` prefix is configable (`asab.Config`, with `/Templates` being default)

**New tests**

-  Render/Download Report
-  Email Body from template
-  Email attachment from template
-  Slack messages

### CODE REVIEW

File: qa.md [hunk 1]
```diff
@@ -33,7 +33,7 @@
 ```
 
 
-## TSM03: Send an email to multiple To, CC and BCC
+## TSM03: Send an email to multiple To, CC and BCC.
```
Comment (mithunbharadwaj) [inline qa.md:36]: @ateska What's the reason behind these numbering. Why there are multiple TSM03's.?

**Comment by mithunbharadwaj** [comment]
@ateska Library folder should be structured based on use case

For emails:
`/Templates/Emails`

For slack:
`/Templates/Slack`

For General purpose:
`/Templates/General`

File: qa.md (hunk 1 above)
Comment (ateska) [reply qa.md]: Discussed yesterday.

File: asabiris/orchestration/sendslack.py [hunk 2]
```diff
@@ -33,7 +33,8 @@ async def send_to_slack(self, msg):
 		# - primarily use absolute path - starts with "/"
 		# - if absolute path is used, check it start with "/Templates"
 		# - if it is not absolute path, it is file name - assume it's a file in Templates folder
-		assert body['template'].startswith("/Templates"), "Template must be stored in /Templates directory"
+		if not body['template'].startswith("/Templates/Slack"):
```
[the remark this answers is not in the document]
Comment (ateska) [reply asabiris/orchestration/sendslack.py:36]: Yes.

**Comment by ateska** [comment]
> @ateska Library folder should be structured based on use case
> 
> For emails: `/Templates/Emails/`
> 
> For slack: `/Templates/Slack/`
> 
> For General purpose: `/Templates/General/`
> 
> If true qa.md should be updated

Yes.

@eliska-n please confirm as well.

**Comment by eliska-n** [comment]
> > @ateska Library folder should be structured based on use case
> > For emails: `/Templates/Emails/`
> > For slack: `/Templates/Slack/`
> > For General purpose: `/Templates/General/`
> > If true qa.md should be updated
> 
> Yes.
> 
> @eliska-n please confirm as well.

 `/Templates/Email/`

**Comment by ateska** [comment]
> > > @ateska Library folder should be structured based on use case
> > > For emails: `/Templates/Emails/`
> > > For slack: `/Templates/Slack/`
> > > For General purpose: `/Templates/General/`
> > > If true qa.md should be updated
> > 
> > 
> > Yes.
> > @eliska-n please confirm as well.
> 
> `/Templates/Email/`

Then `/Templates/Email/`

File: library/Templates/General/hello.html [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE html>
+<html lang="EN">
```
Comment (ateska) [inline library/Templates/General/hello.html:2]: I suggest to include also "subject" for HTML templates.

**Review by ateska — CHANGES_REQUESTED** [review]

File: asabiris/handlers/webhandler.py [hunk 4]
```diff
@@ -176,11 +182,15 @@ async def render(self, request):
...
+		try:
+			html = await self.App.RenderReportOrchestrator.render(template, template_data)
+		except ValueError as e:
+			raise aiohttp.web.HTTPNotFound(text="{}".format(e))
+
 		# get pdf from html if present.
 		if fmt == 'pdf':
 			content_type = "application/pdf"
-			pdf = self.App.HtmlToPdfService.format(html)
+			pdf = self.App.PdfFormatterService.format(html)
```
Comment (ateska) [inline asabiris/handlers/webhandler.py:193]: `self.App.PdfFormatterService.format_html()` ... to enable eventual other formats

**Review by ateska — APPROVED** [review]

**Review by ateska — CHANGES_REQUESTED** [review]
...

File: asabiris/handlers/webhandler.py [hunk 5]
```diff
@@ -55,24 +55,24 @@ async def send_email(self, request, *, json_data):
...
 		"attachments": [
 			{
-			"template": "test.md",
+			"template": "/Templates/Emails/hello.html",
 			"params": {
 				"Name": "Michael Corleone"
 				},
 			"format": "pdf",
-			"filename": "Made.pdf"
+			"filename": "dd-mm--yy.pdf"
```
Comment (ateska) [inline asabiris/handlers/webhandler.py:70]: This is rather misleading name.

**Review by ateska — CHANGES_REQUESTED** [review]

**Comment by mithunbharadwaj** [comment]
@ateska can you review once again , esp can you review qa.md once?

File: qa.md [hunk 6]
```diff
@@ -62,21 +205,122 @@
...
     "body": {
-        "template": "/Templates/hello.html"
+        "template": "/Templates/Emails/hello.html"
     },
-    'attachments': [{
-        'base64': '',
-        'content-type': 'text/csv',
-        'filename': 'c7fbbb3d716d4d7c95d3b887b288ed62.csv'
+    "attachments": [{
+        "base64": "",
```
Comment (ateska) [inline qa.md:231]: There should be something here, small Base64 CSV

**Review by ateska — CHANGES_REQUESTED** [review]

File: qa.md (hunk 6 above)
Comment (mithunbharadwaj) [reply qa.md]: @ateska Can I merge this?

File: asabiris/handlers/webhandler.py [hunk 7]
```diff
@@ -104,6 +104,9 @@ async def send_email(self, request, *, json_data):
 				attachments=json_data.get("attachments", []),  # Optional
 			)
 
+		except ValueError as e:
```
Comment (ateska) [inline asabiris/handlers/webhandler.py:107]: Why ValueError is considered as "HTTPNotFound"?
That is behind this error?

File: asabiris/handlers/webhandler.py [hunk 8]
```diff
@@ -143,6 +146,9 @@ async def send_alert(self, request, *, json_data):
 		except jinja2.exceptions.UndefinedError as e:
 			raise aiohttp.web.HTTPBadRequest(text="Jinja2 error: {}".format(e))
 
+		except ValueError as e:
```
Comment (ateska) [inline asabiris/handlers/webhandler.py:149]: dtto

File: asabiris/handlers/webhandler.py (hunk 4 above)
Comment (ateska) [reply asabiris/handlers/webhandler.py:193]: This comment is not resolved.

File: library/Templates/General/hello.html (hunk 3 above)
Comment (ateska) [reply library/Templates/General/hello.html:2]: Not yet resolved.

**Review by ateska — CHANGES_REQUESTED** [review]

File: asabiris/handlers/webhandler.py (hunk 4 above)
Comment (mithunbharadwaj) [reply asabiris/handlers/webhandler.py:193]: @ateska  I thought it made sense to solve this when we introduce when we deal with other formats. Is that okay?

File: library/Templates/General/hello.html (hunk 3 above)
Comment (mithunbharadwaj) [reply library/Templates/General/hello.html:2]: @ateska Let me know if you are happy with the HTML template

File: asabiris/handlers/webhandler.py (hunk 8 above)
Comment (mithunbharadwaj) [reply asabiris/handlers/webhandler.py:149]: https://github.com/TeskaLabs/asab-iris/pull/14/files#diff-7408af1841a6261bcfc012878c4c6c067eb50a1e0a4dbc978db20a90f3653211R74

I am raising value error here.

**Comment by mithunbharadwaj** [comment]
@ateska

I would like to work this on next MR.

` /Templates prefix is configable (asab.Config, with /Templates being default) `

File: asabiris/handlers/webhandler.py (hunk 7 above)
Comment (ateska) [reply asabiris/handlers/webhandler.py:107]: Then don't catch this here ... just pass that up. It's OK to produce 501 to the user in this situation.

File: asabiris/handlers/webhandler.py (hunk 8 above)
Comment (ateska) [reply asabiris/handlers/webhandler.py:149]: dtto

File: qa.md (hunk 6 above)
Comment (ateska) [reply qa.md]: Can you actually include the base64 encoded small CSV file here?

**Review by ateska — CHANGES_REQUESTED** [review]

**Comment by ateska** [comment]
> @ateska
> 
> I would like to work this on next MR.
> 
> `/Templates prefix is configable (asab.Config, with /Templates being default)`

Yes.

File: library/Templates/General/hello.html (hunk 3 above)
Comment (ateska) [reply library/Templates/General/hello.html:2]: Do you have example somewhere with the Subject being part of the HTML template?

File: asabiris/handlers/webhandler.py (hunk 4 above)
Comment (ateska) [reply asabiris/handlers/webhandler.py:193]: hmms, ok then

**Review by ateska — CHANGES_REQUESTED** [review]

File: qa.md (hunk 6 above)
Comment (mithunbharadwaj) [reply qa.md]: @ateska It was here.

https://github.com/TeskaLabs/asab-iris/pull/14/files#diff-6ffbbeade89893ca3b4a89dc9c06569fc1b77827c669f0493a92e1c2c716e152R253

File: library/Templates/General/hello.html (hunk 3 above)
Comment (mithunbharadwaj) [reply library/Templates/General/hello.html:2]: @ateska `def find_subject_in_html(body):
	regex = r"(<title>(.*)</title>)"
	match = re.search(regex, body)
	if match is None:
		return body, None
	subject_line, subject = match.groups()
	body = body.replace(subject_line, "")
	return body, subject
`

I think title is assumed to be subject , correct?

https://github.com/TeskaLabs/asab-iris/blob/5e32bffe0093386fc10bde6b3843eec2740db580/library/Templates/Email/hello.html#L3

File: qa.md (hunk 6 above)
Comment (ateska) [reply qa.md]: OK

File: library/Templates/General/hello.html (hunk 3 above)
Comment (ateska) [reply library/Templates/General/hello.html:2]: Yes, remove `-` and it is correct.

**Review by ateska — APPROVED** [review]

**Comment by ateska** [comment]
Create release tag `v22.11-alpha6` after merging, please. @mithunbharadwaj

**Pull request merged**